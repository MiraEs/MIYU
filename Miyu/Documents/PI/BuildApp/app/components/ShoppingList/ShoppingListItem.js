import React, {
	Component,
} from 'react';
import PropTypes from 'prop-types';
import {
	StyleSheet,
	TouchableOpacity,
	View,
	ViewPropTypes,
} from 'react-native';
import {
	Image,
	Text,
} from 'BuildLibrary';
import { withNavigation } from '@expo/ex-navigation';
import styles from '../../lib/styles';
import Icon from 'react-native-vector-icons/Ionicons';
import helpers from '../../lib/helpers';
import {
	PRODUCT_SECTION,
	IMAGE_42,
} from '../../constants/CloudinaryConstants';
import {
	HOME,
	MAIN,
} from '../../constants/constants';
import {
	addItemsToCartFromProject,
	removeItemFromShoppingList,
} from '../../actions/ProjectActions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Swipeable from 'react-native-swipeable';
import SwipeRowAction from '../SwipeRowAction';
import { showAlert } from '../../actions/AlertActions';
import PricedOptions from '../Cart/PricedOptions';
import router from '../../router';
import TrackingActions from '../../lib/analytics/TrackingActions';
import { trackAction } from '../../actions/AnalyticsActions';

const componentStyles = StyleSheet.create({
	purchased: {
		flexDirection: 'row',
		marginBottom: styles.measurements.gridSpace1,
		marginRight: styles.measurements.gridSpace2,
		alignItems: 'center',
	},
	itemWrapper: {
		marginTop: -1,
		borderColor: styles.colors.grey,
		borderWidth: styles.dimensions.borderWidth,
	},
	subItem: {
		marginLeft: styles.measurements.gridSpace2,
		borderLeftWidth: styles.dimensions.borderWidthLarge,
		borderLeftColor: styles.colors.greyDark,
	},
	imageWrapper: {
		flexDirection: 'row',
		margin: styles.measurements.gridSpace1,
		alignItems: 'center',
	},
	image: {
		marginLeft: styles.measurements.gridSpace1,
		marginRight: styles.measurements.gridSpace2,
	},
	secondRowWrapper: {
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center',
	},
	editButton: {
		paddingLeft: styles.measurements.gridSpace2,
	},
	purchaseIcon: {
		marginRight: styles.measurements.gridSpace1,
	},
	pricedOptions: {
		marginLeft: 70,
		marginBottom: 3,
	},
});

export class ShoppingListItem extends Component {

	constructor(props) {
		super(props);
		this.state = {
			addToCartEnabled: true,
		};
		this.swipeable = {
			recenter: helpers.noop,
		};
	}

	getBackgroundStyle = () => {
		const { item, isSubItem } = this.props;
		const backgroundColor = item.isPurchased || isSubItem ? styles.colors.greyLight : styles.colors.white;
		return {
			backgroundColor,
		};
	};

	getStyle = () => {
		return StyleSheet.flatten([
			componentStyles.itemWrapper,
			this.props.isSubItem ? componentStyles.subItem : {},
			this.getBackgroundStyle(),
			this.props.style,
		]);
	};

	onPressEdit = () => {
		const {
			item,
			navigation,
			onUpdateItem,
			projectId,
			sessionCartId,
		} = this.props;
		this.swipeable.recenter();
		navigation.getNavigator('root').push('productEdit', {
			item,
			onUpdateItem,
			projectId,
			sessionCartId,
		});
	};

	onPressAddToCart = () => {
		this.setState({ addToCartEnabled: false });
		const { item, actions, projectId } = this.props;
		actions.addItemsToCartFromProject(projectId, [item])
			.then(() => this.swipeable.recenter())
			.catch(() => this.props.actions.showAlert('Failed to add item to cart.', 'error'))
			.done();
	};

	onPressDeleteItem = () => {
		const { item: { itemKey, sessionCartId }, projectId } = this.props;
		this.props.actions.removeItemFromShoppingList(sessionCartId, itemKey, projectId)
			.catch(() => this.props.actions.showAlert('Failed to delete item.', 'error'))
			.done();
	};

	goToProduct = () => {
		const { item: { product }} = this.props;
		this.props.navigation.getNavigatorByUID(MAIN).jumpToTab(HOME);
		requestAnimationFrame(() => this.props.navigation.performAction(({ stacks }) => {
			stacks(HOME).push(router.getRoute('productDetail', {
				compositeId: product.productCompositeId,
				uniqueId: product.uniqueId,
				manufacturer: product.manufacturer,
				sku: product.sku,
				finish: product.finish,
			}));
		}));
	};

	getSwipeRowAction = (options) => {
		const { text, backgroundColor, iconName, analyticsData, onPress, rightSwipeItem = true } = options;
		return (
			<SwipeRowAction
				text={text}
				iconName={iconName}
				backgroundColor={backgroundColor}
				analyticsData={analyticsData}
				rightSwipeItem={rightSwipeItem}
				onPress={onPress}
			/>
		);
	};

	getAllSwipeActions = () => {
		const swipeActions = [];
		const { isPurchased, product } = this.props.item;
		if (!isPurchased) {
			swipeActions.push(
				this.getSwipeRowAction({
					text: 'Add To Cart',
					backgroundColor: styles.colors.primary,
					iconName: 'cart',
					analyticsData: {
						trackName: TrackingActions.SHOPPING_LIST_ITEM_ADD_TO_CART_TAP,
						trackData: {
							screen: 'Shopping List Tab',
							productId: product.productId,
							manufacturer: product.manufacturer,
							uniqueId: product.uniqueId,
							compositeId: product.productCompositeId,
							finish: product.finish,
						},
					},
					onPress: this.onPressAddToCart,
				})
			);
		}
		swipeActions.push(
			this.getSwipeRowAction({
				text: 'Edit',
				backgroundColor: styles.colors.accent,
				iconName: 'create',
				analyticsData: { trackName: TrackingActions.SHOPPING_LIST_ITEM_EDIT_TAP },
				onPress: this.onPressEdit,
			})
		);
		if (!isPurchased) {
			swipeActions.push(
				this.getSwipeRowAction({
					text: 'Delete',
					backgroundColor: styles.colors.error,
					iconName: 'trash',
					analyticsData: { trackName: TrackingActions.SHOPPING_LIST_ITEM_DELETE_TAP },
					onPress: this.onPressDeleteItem,
				})
			);
		}
		return swipeActions;
	};

	renderPurchasedTag = () => {
		const { item } = this.props;
		if (item.isPurchased) {
			return (
				<View style={componentStyles.purchased}>
					<Icon
						name={helpers.getIcon('checkmark-circle')}
						size={18}
						color={styles.colors.primary}
						style={componentStyles.purchaseIcon}
					/>
					<Text lineHeight={false}>Purchased</Text>
				</View>
			);
		}
	};

	renderTotalPriceAndPurchase = () => {
		const { item: {
			isPurchased,
			quantityPurchased,
			quantityUnpurchased,
			unitPrice,
		}} = this.props;
		const quantity = isPurchased ? quantityPurchased : quantityUnpurchased;
		const itemSubTotal = unitPrice * quantity;
		return (
			<View style={componentStyles.secondRowWrapper}>
				<Text color="primary">
					{helpers.toUSD(itemSubTotal)}
				</Text>
				{this.renderPurchasedTag()}
			</View>
		);
	};

	renderProductDetails = () => {
		const { item } = this.props;
		const {
			estimatedArrival = 'Arriving Monday',
			isPurchased,
			orderNum = '123456',
			product = {},
			quantityPurchased,
			quantityUnpurchased,
			unitPrice,
		} = item;

		if (isPurchased) {
			return (
				<View style={styles.elements.flex1}>
					<Text weight="bold">
						{product.displayName}
					</Text>
					<Text
						color="primary"
						decoration="underline"
					>
						Order #{orderNum}
					</Text>
					<Text size="small">(Qty. {quantityPurchased}) {estimatedArrival}</Text>
					{this.renderTotalPriceAndPurchase()}
				</View>
			);
		}

		return (
			<View>
				<Text weight="bold">
					{product.displayName}
				</Text>
				<Text>Finish: {product.finish}</Text>
				<View style={styles.elements.flexRow}>
					<Text size="small">{helpers.toUSD(unitPrice)}{' '}</Text>
					<Text size="small">(Qty. {quantityUnpurchased}){' '}</Text>
					<Text
						size="small"
						color="greyDark"
					>
						{helpers.toBigNumber(product.stockCount)} In-stock
					</Text>
				</View>
				{this.renderTotalPriceAndPurchase()}
			</View>
		);
	};

	render() {
		const { item, bounceOnMount, actions } = this.props;
		const { product = {} } = item;
		const imageUri = helpers.getCloudinaryImageUrl({
			...IMAGE_42,
			section: PRODUCT_SECTION,
			manufacturer: product.manufacturer,
			name: product.image,
		});

		return (
			<Swipeable
				onRef={(ref) => (this.swipeable = ref ? ref : this.swipeable)}
				rightButtons={this.getAllSwipeActions()}
				bounceOnMount={bounceOnMount}
				onRightActionRelease={() => actions.trackAction(TrackingActions.SHOPPING_LIST_ITEM_SWIPE)}
			>
				<View style={this.getStyle()}>
					<View style={componentStyles.imageWrapper}>
						<TouchableOpacity onPress={this.goToProduct}>
							<Image
								source={{ uri: imageUri }}
								height={IMAGE_42.height}
								width={IMAGE_42.width}
								style={componentStyles.image}
							/>
						</TouchableOpacity>
						{this.renderProductDetails()}
					</View>
					<View style={componentStyles.pricedOptions}>
						<PricedOptions options={item.pricedOptions}/>
					</View>
				</View>
			</Swipeable>
		);
	}
}

ShoppingListItem.propTypes = {
	actions: PropTypes.object,
	bounceOnMount: PropTypes.bool,
	item: PropTypes.object,
	isSubItem: PropTypes.bool,
	navigation: PropTypes.shape({
		getNavigator: PropTypes.func,
		getNavigatorByUID: PropTypes.func,
		performAction: PropTypes.func,
	}),
	onUpdateItem: PropTypes.func,
	projectId: PropTypes.number.isRequired,
	sessionCartId: PropTypes.number.isRequired,
	style: ViewPropTypes.style,
};

ShoppingListItem.defaultProps = {
	bounceOnMount: false,
	item: {},
	isSubItem: false,
	onUpdateItem: helpers.noop,
};

const mapDispatchToProps = (dispatch) => {
	return {
		actions: bindActionCreators({
			addItemsToCartFromProject,
			removeItemFromShoppingList,
			showAlert,
			trackAction,
		}, dispatch),
	};
};

export default withNavigation(connect(null, mapDispatchToProps)(ShoppingListItem));
