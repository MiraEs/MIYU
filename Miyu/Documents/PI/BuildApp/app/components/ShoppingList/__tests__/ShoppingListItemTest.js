jest.mock('react-native');
jest.mock('BuildNative');
jest.mock('BuildLibrary');
jest.mock('react-native-vector-icons/Ionicons', () => 'Icon');
jest.mock('react-native-swipeable', () => 'Swipeable');
jest.mock('../../SwipeRowAction', () => 'SwipeRowAction');
jest.mock('../../SwipeRowAction', () => 'SwipeRowAction');
jest.mock('../../Cart/PricedOptions', () => 'PricedOptions');
jest.mock('../../../router', () => ({
	getRoute: jest.fn(),
}));


import { ShoppingListItem } from '../ShoppingListItem';
import React from 'react';
import renderer from 'react-test-renderer';

const defaultProps = {
	item: {
		isPurchased: true,
		name: 'test',
		quantityPurchased: 1,
		quantityUnpurchased: 1,
		estimatedArrival: 'Arriving Monday',
		orderNum: '123456',
		product: {
			displayName: 'test',
			finish: 'steel',
			image: 'test',
			manufacturer: 'test',
			stockCount: 1,
		},
		unitPrice: 1,
	},
	navigation: {
		getNavigator: jest.fn(() => ({
			push: jest.fn(),
		})),
		getNavigatorByUID: jest.fn(),
		performAction: jest.fn(),
	},
	projectId: 1,
	sessionCartId: 2,
};

describe('ShoppingListItem component', () => {
	it('it should render', () => {
		const tree = renderer.create(<ShoppingListItem {...defaultProps} />);
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it('it should render collapsed', () => {
		const newProps = { ...defaultProps };
		newProps.item.isPurchased = false;
		const tree = renderer.create(
			<ShoppingListItem
				{...newProps}
			/>
		);
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it('it should render a child item', () => {
		const tree = renderer.create(
			<ShoppingListItem
				{...defaultProps}
				isSubItem={true}
			/>
		);
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it('should push productEdit to navigator', () => {
		const instance = new ShoppingListItem(defaultProps);
		instance.onPressEdit();
		expect(instance.props.navigation.getNavigator).toHaveBeenCalled();
	});

});
