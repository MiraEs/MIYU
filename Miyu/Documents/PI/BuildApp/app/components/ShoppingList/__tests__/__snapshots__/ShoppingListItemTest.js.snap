// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingListItem component it should render 1`] = `
<Swipeable
  bounceOnMount={false}
  onRef={[Function]}
  onRightActionRelease={[Function]}
  rightButtons={
    Array [
      <SwipeRowAction
        analyticsData={
            Object {
                "trackName": "build:app:projects:details:shoppinglist:edititemtap",
              }
        }
        backgroundColor="#D45D00"
        iconName="create"
        onPress={[Function]}
        rightSwipeItem={true}
        text="Edit"
    />,
    ]
  }
>
  <View
    style={
      Object {
        "0": Object {
          "borderColor": "#D0D3D4",
          "borderWidth": 1,
          "marginTop": -1,
        },
        "1": Object {},
        "2": Object {
          "backgroundColor": "#EBEDEF",
        },
        "3": undefined,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 8,
        }
      }
    >
      <TouchableOpacity
        onPress={[Function]}
      >
        <Image
          height={42}
          source={
            Object {
              "uri": "http://s3.img-b.com/image/private/t_base,f_auto,dpr_2.0,c_lpad,w_42,h_42/product/test/test",
            }
          }
          style={
            Object {
              "marginLeft": 8,
              "marginRight": 16,
            }
          }
          width={42}
        />
      </TouchableOpacity>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          weight="bold"
        >
          test
        </Text>
        <Text
          color="primary"
          decoration="underline"
        >
          Order #
          123456
        </Text>
        <Text
          size="small"
        >
          (Qty. 
          1
          ) 
          Arriving Monday
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            color="primary"
          >
            $1.00
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "marginBottom": 8,
                "marginRight": 16,
              }
            }
          >
            <Icon
              color="#00A499"
              name="ios-checkmark-circle"
              size={18}
              style={
                Object {
                  "marginRight": 8,
                }
              }
            />
            <Text
              lineHeight={false}
            >
              Purchased
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "marginBottom": 3,
          "marginLeft": 70,
        }
      }
    >
      <PricedOptions
        options={undefined}
      />
    </View>
  </View>
</Swipeable>
`;

exports[`ShoppingListItem component it should render a child item 1`] = `
<Swipeable
  bounceOnMount={false}
  onRef={[Function]}
  onRightActionRelease={[Function]}
  rightButtons={
    Array [
      <SwipeRowAction
        analyticsData={
            Object {
                "trackData": Object {
                  "compositeId": undefined,
                  "finish": "steel",
                  "manufacturer": "test",
                  "productId": undefined,
                  "screen": "Shopping List Tab",
                  "uniqueId": undefined,
                },
                "trackName": "build:app:projects:details:shoppinglist:additemtocarttap",
              }
        }
        backgroundColor="#00A499"
        iconName="cart"
        onPress={[Function]}
        rightSwipeItem={true}
        text="Add To Cart"
    />,
      <SwipeRowAction
        analyticsData={
            Object {
                "trackName": "build:app:projects:details:shoppinglist:edititemtap",
              }
        }
        backgroundColor="#D45D00"
        iconName="create"
        onPress={[Function]}
        rightSwipeItem={true}
        text="Edit"
    />,
      <SwipeRowAction
        analyticsData={
            Object {
                "trackName": "build:app:projects:details:shoppinglist:deleteitemtap",
              }
        }
        backgroundColor="#A94442"
        iconName="trash"
        onPress={[Function]}
        rightSwipeItem={true}
        text="Delete"
    />,
    ]
  }
>
  <View
    style={
      Object {
        "0": Object {
          "borderColor": "#D0D3D4",
          "borderWidth": 1,
          "marginTop": -1,
        },
        "1": Object {
          "borderLeftColor": "#B1B2B4",
          "borderLeftWidth": 2,
          "marginLeft": 16,
        },
        "2": Object {
          "backgroundColor": "#EBEDEF",
        },
        "3": undefined,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 8,
        }
      }
    >
      <TouchableOpacity
        onPress={[Function]}
      >
        <Image
          height={42}
          source={
            Object {
              "uri": "http://s3.img-b.com/image/private/t_base,f_auto,dpr_2.0,c_lpad,w_42,h_42/product/test/test",
            }
          }
          style={
            Object {
              "marginLeft": 8,
              "marginRight": 16,
            }
          }
          width={42}
        />
      </TouchableOpacity>
      <View>
        <Text
          weight="bold"
        >
          test
        </Text>
        <Text>
          Finish: 
          steel
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            size="small"
          >
            $1.00
             
          </Text>
          <Text
            size="small"
          >
            (Qty. 
            1
            )
             
          </Text>
          <Text
            color="greyDark"
            size="small"
          >
            1
             In-stock
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            color="primary"
          >
            $1.00
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "marginBottom": 3,
          "marginLeft": 70,
        }
      }
    >
      <PricedOptions
        options={undefined}
      />
    </View>
  </View>
</Swipeable>
`;

exports[`ShoppingListItem component it should render collapsed 1`] = `
<Swipeable
  bounceOnMount={false}
  onRef={[Function]}
  onRightActionRelease={[Function]}
  rightButtons={
    Array [
      <SwipeRowAction
        analyticsData={
            Object {
                "trackData": Object {
                  "compositeId": undefined,
                  "finish": "steel",
                  "manufacturer": "test",
                  "productId": undefined,
                  "screen": "Shopping List Tab",
                  "uniqueId": undefined,
                },
                "trackName": "build:app:projects:details:shoppinglist:additemtocarttap",
              }
        }
        backgroundColor="#00A499"
        iconName="cart"
        onPress={[Function]}
        rightSwipeItem={true}
        text="Add To Cart"
    />,
      <SwipeRowAction
        analyticsData={
            Object {
                "trackName": "build:app:projects:details:shoppinglist:edititemtap",
              }
        }
        backgroundColor="#D45D00"
        iconName="create"
        onPress={[Function]}
        rightSwipeItem={true}
        text="Edit"
    />,
      <SwipeRowAction
        analyticsData={
            Object {
                "trackName": "build:app:projects:details:shoppinglist:deleteitemtap",
              }
        }
        backgroundColor="#A94442"
        iconName="trash"
        onPress={[Function]}
        rightSwipeItem={true}
        text="Delete"
    />,
    ]
  }
>
  <View
    style={
      Object {
        "0": Object {
          "borderColor": "#D0D3D4",
          "borderWidth": 1,
          "marginTop": -1,
        },
        "1": Object {},
        "2": Object {
          "backgroundColor": "#FFFFFF",
        },
        "3": undefined,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 8,
        }
      }
    >
      <TouchableOpacity
        onPress={[Function]}
      >
        <Image
          height={42}
          source={
            Object {
              "uri": "http://s3.img-b.com/image/private/t_base,f_auto,dpr_2.0,c_lpad,w_42,h_42/product/test/test",
            }
          }
          style={
            Object {
              "marginLeft": 8,
              "marginRight": 16,
            }
          }
          width={42}
        />
      </TouchableOpacity>
      <View>
        <Text
          weight="bold"
        >
          test
        </Text>
        <Text>
          Finish: 
          steel
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            size="small"
          >
            $1.00
             
          </Text>
          <Text
            size="small"
          >
            (Qty. 
            1
            )
             
          </Text>
          <Text
            color="greyDark"
            size="small"
          >
            1
             In-stock
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            color="primary"
          >
            $1.00
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "marginBottom": 3,
          "marginLeft": 70,
        }
      }
    >
      <PricedOptions
        options={undefined}
      />
    </View>
  </View>
</Swipeable>
`;
